Convert UK Benefits Abroad flow tests

leenagupte opened the pull request:
Trello: https://trello.com/c/0hOseszb

# What's changed and why?

Converts the UK Benefits Abroad flow tests to use the node based testing approach introduced in [ADR-3][1].

Also removes the existing flow integration tests as they are no longer needed.

[1]: https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+        end
+
+        should "have a next node of is_any_of_the_following_apply? for an empty response" do
+          assert_next_node :is_any_of_the_following_apply?, for_response: ""
```
minor aside; 😬  the grammar of that node name - unless `is` is an abbreviation

leenagupte replied:
😬

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+          assert_next_node :wfp_going_abroad_eea_maybe_outcome, for_response: "before_jan_2021"
+        end
+
+        should "have a next node of esa_going_abroad_eea_outcome for a 'before_jan_2021' response if benefit is esa if going_abroad" do
```
```suggestion
        should "have a next node of esa_going_abroad_eea_outcome for a 'before_jan_2021' response if benefit is esa and going_abroad" do
```

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+          assert_next_node :esa_going_abroad_eea_outcome, for_response: "before_jan_2021"
+        end
+
+        should "have a next node of esa_already_abroad_eea_outcome for a 'before_jan_2021' response if benefit is esa if already_abroad" do
```
```suggestion
        should "have a next node of esa_already_abroad_eea_outcome for a 'before_jan_2021' response if benefit is esa and already_abroad" do
```

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+          assert_next_node :esa_already_abroad_eea_outcome, for_response: "before_jan_2021"
+        end
+
+        should "have a next node of db_going_abroad_eea_outcome for a 'before_jan_2021' response if benefit is disability_benefits if going_abroad" do
```
```suggestion
        should "have a next node of db_going_abroad_eea_outcome for a 'before_jan_2021' response if benefit is disability_benefits and going_abroad" do
```

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+          assert_next_node :db_going_abroad_eea_outcome, for_response: "before_jan_2021"
+        end
+
+        should "have a next node of db_already_abroad_eea_outcome for a 'before_jan_2021' response if benefit is disability_benefits if already_abroad" do
```
```suggestion
        should "have a next node of db_already_abroad_eea_outcome for a 'before_jan_2021' response if benefit is disability_benefits and already_abroad" do
```

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+
+        should "have a next node of jsa_not_entitled_outcome for a 'no' response if benefit is jsa" do
+          assert_next_node :jsa_not_entitled_outcome, for_response: "no"
+        end
```
It might be a bit neater to have a loop of `%w[after_jan_2021 no]` responses to save duplicating each of the 6 outcomes.

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+        assert_rendered_outcome text: "You can carry on getting tax credits for up to 12 weeks after leaving the UK if you were already receiving them before you left."
+      end
+    end
+  end
```
Great job going through all these many outcomes to find the conditionals!

kevindew reviewed (commented):
This looks great ⭐  well done working through such a massive Smart Answer.

Most of my comments are pretty minor, only recurring one is the former Yugoslavia thing. There were more cases of that than I tagged but we can deal with that once we resolve the concern.

leenagupte commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+
+          should "have next node of iidb_going_abroad_eea_outcome for any response in EEA but not in former Yugoslavia if going_abroad" do
+            assert_next_node :iidb_going_abroad_eea_outcome, for_response: "liechtenstein"
+          end
```
Have tests for those. Lines 222 and 231:

`should "have next node of iidb_going_abroad_ss_outcome for any response not in EEA but in former Yugoslavia if going_abroad" do`

`should "have next node of iidb_already_abroad_ss_outcome for any response not in EEA but in former Yugoslavia if already_abroad" do`

leenagupte replied:
I'll give it a go. Have you got a country recommendation? 😬

leenagupte replied:
Perhaps Russia?

kevindew replied:
Ooh looks like you've sorted that now so I don't think you need a recommendation right?

leenagupte commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1328 @@
+        end
+
+        should "have a next node of child_benefit_not_entitled_outcome for an empty response" do
+          assert_next_node :child_benefit_not_entitled_outcome, for_response: ""
```
No, this is a checkkbox question, so it's the equivalent of not selecting anything at all.

leenagupte commented:
@kevindew Thanks for your review. I've fixed up most of your suggestions. I've added a temporary commit with suggested changes to the "former Yugoslavia" to be more explicit about which list of countries we're expecting as if "former Yugoslavia" is too narrow, then "social security country" is too broad.

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1302 @@
+    end
+  end
+
+  context "question: which_benefit" do
```
```suggestion
  context "question: which_benefit?" do
```

Teeny consistency issue

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1302 @@
+                          iidb_already_claiming?: "yes"
+          end
+
+          should "have next node of iidb_going_abroad_eea_outcome for any response in EEA but not in social security country with iidb if going_abroad" do
```
```suggestion
          should "have next node of iidb_going_abroad_eea_outcome for any response in an EEA country if going_abroad" do
```

Since the code is an OR it shouldn't matter whether the country is in the social-security-country-with-iidb-group or not

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1302 @@
+
+          should "have next node of iidb_going_abroad_ss_outcome for any response not in EEA but in social security country with iidb if going_abroad" do
+            assert_next_node :iidb_going_abroad_ss_outcome, for_response: "north-macedonia"
+          end
```
Looks good 👍

kevindew commented on test/flows/uk_benefits_abroad_flow_test.rb:
```diff
@@ -0,0 +1,1302 @@
+            assert_next_node :iidb_going_abroad_eea_outcome, for_response: "australia"
+          end
+
+          should "have next node of iidb_already_abroad_eea_outcome for any response in EEA country but not in social security country with iidb if already_abroad" do
```
Same point as above about the OR negating the not social security country

kevindew reviewed (commented):
I like what you did with the Former Yugoslavia problem, I know they're a bit of a mouthful in the tests (and verbal expression of them is questionable) but it's great that they match the calculator code (and you'd hope any future improvements involved both the tests and the calculator code improving)

leenagupte commented:
@kevindew Thanks for the re-review. I've fixed the last of the niggles and fixed up the commits.

kevindew reviewed (approved):
Nice one
